Lazy field definitions not working for Input types

[Release v14.7.0](https://github.com/webonyx/graphql-php/releases/tag/v14.7.0) allows providing field definitions as a callable and resolve them lazily, thanks to the great work by @stayallive in PR #871 

I was trying to port our codebase so that fields are resolved lazily, and noticed that this feature does not work for input object types.

@stayallive [suggested](https://github.com/webonyx/graphql-php/pull/861#issuecomment-892093535) I open an issue.

You can add this test below to `DefinitionTest.php` to reproduce the issue:

```php
    /**
     * @see it('allows an input type to define its fields as closure returning array field definition to be lazy loaded')
     */
    public function testAllowsInputTypeWhichDefinesItFieldsAsClosureReturningFieldDefinitionAsArray() : void
    {
        $objType = new InputObjectType([
            'name'   => 'SomeInputObject',
            'fields' => [
                // 'f' => ['type' => Type::string()],
                'f' => static function () : array {
                    return ['type' => Type::string()];
                },
            ],
        ]);

        $objType->assertValid();

        self::assertSame(Type::string(), $objType->getField('f')->getType());
    }
```

This test will fail. 
```
./vendor/bin/phpunit --filter testAllowsInputTypeWhichDefinesItFieldsAsClosureReturningFieldDefinitionAsArray
PHPUnit 8.5.19 by Sebastian Bergmann and contributors.

E                                                                   1 / 1 (100%)

Time: 250 ms, Memory: 14.00 MB

There was 1 error:

1) GraphQL\Tests\Type\DefinitionTest::testAllowsInputTypeWhichDefinesItFieldsAsClosureReturningFieldDefinitionAsArray
TypeError: Return value of GraphQL\Type\Schema::resolveType() must be an instance of GraphQL\Type\Definition\Type, array returned

graphql/graphql-php/src/Type/Schema.php:406
graphql/graphql-php/src/Type/Definition/InputObjectField.php:121
graphql/graphql-php/src/Type/Definition/InputObjectField.php:148
graphql/graphql-php/src/Type/Definition/InputObjectType.php:117
graphql/graphql-php/tests/Type/DefinitionTest.php:1920

ERRORS!
Tests: 1, Assertions: 0, Errors: 1.
```

The code in the comment `// 'f' => ['type' => Type::string()],` will pass, since it is not using a callable.

This is a promising feature, I can see some gains already, but we have a lot of input types that could be converted, it would be great if this was supported for them as well.

Thanks a lot.

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -26,6 +26,7 @@ You can find and compare releases at the [GitHub release page](https://github.co
 - Add tests for errors that occur when undeclared fields are passed in input
 - Warn about orphaned object types
 - Expose structured enumeration of directive locations
+- Allow lazy input object fields
 
 ### Optimized
 

--- a/src/Type/Definition/InputObjectType.php
+++ b/src/Type/Definition/InputObjectType.php
@@ -32,7 +32,7 @@ class InputObjectType extends Type implements InputType, NullableType, NamedType
     public array $extensionASTNodes;
 
     /**
-     * @param mixed[] $config
+     * @param array<string, mixed> $config
      */
     public function __construct(array $config)
     {
@@ -60,16 +60,21 @@ class InputObjectType extends Type implements InputType, NullableType, NamedType
 
         Utils::invariant(isset($this->fields[$name]), "Field '%s' is not defined for type '%s'", $name, $this->name);
 
+        $this->initializeField($name);
+
         return $this->fields[$name];
     }
 
     /**
-     * @return InputObjectField[]
+     * @return array<string, InputObjectField>
      */
     public function getFields(): array
     {
         if (! isset($this->fields)) {
             $this->initializeFields();
+            foreach ($this->fields as $name => $_) {
+                $this->initializeField($name);
+            }
         }
 
         return $this->fields;
@@ -77,8 +82,7 @@ class InputObjectType extends Type implements InputType, NullableType, NamedType
 
     protected function initializeFields(): void
     {
-        $this->fields = [];
-        $fields       = $this->config['fields'] ?? [];
+        $fields = $this->config['fields'] ?? [];
         if (is_callable($fields)) {
             $fields = $fields();
         }
@@ -89,14 +93,26 @@ class InputObjectType extends Type implements InputType, NullableType, NamedType
             );
         }
 
-        foreach ($fields as $name => $field) {
-            if ($field instanceof Type || is_callable($field)) {
-                $field = ['type' => $field];
-            }
+        $this->fields = $fields;
+    }
 
-            $field                      = new InputObjectField($field + ['name' => $name]);
-            $this->fields[$field->name] = $field;
+    protected function initializeField(string $name): void
+    {
+        $field = $this->fields[$name];
+        if ($field instanceof InputObjectField) {
+            return;
         }
+
+        if (is_callable($field)) {
+            $field = $field();
+        }
+
+        if ($field instanceof Type || is_callable($field)) {
+            $field = ['type' => $field];
+        }
+
+        $field               = new InputObjectField($field + ['name' => $name]);
+        $this->fields[$name] = $field;
     }
 
     /**

--- a/tests/Type/DefinitionTest.php
+++ b/tests/Type/DefinitionTest.php
@@ -750,6 +750,22 @@ class DefinitionTest extends TestCase
         self::assertSame($someMutation->getField('mutateSomething')->getArg('input')->getType(), $inputObject);
     }
 
+    public function testAllowsInputTypeWhichDefinesItFieldsAsClosureReturningFieldDefinitionAsArray(): void
+    {
+        $objType = new InputObjectType([
+            'name'   => 'SomeInputObject',
+            'fields' => [
+                'f' => static fn (): array => [
+                    'type' => Type::string(),
+                ],
+            ],
+        ]);
+
+        $objType->assertValid();
+
+        self::assertSame(Type::string(), $objType->getField('f')->getType());
+    }
+
     public function testInterfaceTypeAllowsRecursiveDefinitions(): void
     {
         $called    = false;

